### PR TITLE
Link settings extensibility

### PR DESCRIPTION
**Pull request by abotteram**
## Description

Adds the possibility to expand the inline link settings via Slot/Fill and alter the link's attributes.

To achieve this I had to refactor the logic that determined the link's attributes because the only possibility was to change the link to an external link.

While working on this I found out that the `aria-label` logic is broken. The label will always be ` (opens in new window)`.

## How has this been tested?

Add the `NoFollowToggleWrapper` somewhere in your code and make sure the component is always rendered. I simply copied it into `packages/format-library/src/inline.js`.

```js
class NoFollowToggle extends Component {
	isChecked() {
		const { rel } = this.props.attributes;

		if ( ! rel ) {
			return false;
		}

		return rel.split( ' ' ).includes( 'nofollow' );
	}

	toggle() {
		const { setLinkAttributes, attributes } = this.props;

		const rel = attributes.rel;

		if ( this.isChecked() ) {
			if ( ! rel ) {
				return;
			}

			const newRel = rel.split( ' ' ).filter( ( relItem ) => {
				return relItem !== 'nofollow';
			} ).join( ' ' );

			setLinkAttributes( {
				...attributes,
				rel: newRel,
			} );
			return;
		}

		if ( ! rel ) {
			setLinkAttributes( {
				...attributes,
				rel: 'nofollow',
			} );
		} else {
			setLinkAttributes( {
				...attributes,
				rel: [ rel, 'nofollow' ].join( ' ' ),
			} );
		}
	}

	render() {
		return (
			<ToggleControl
				label={ 'No Follow' }
				checked={ this.isChecked() }
				onChange={ this.toggle.bind( this ) }
			/>
		);
	}
}

function NoFollowToggleWrapper() {
	return (
		<Fill name="LinkSettings">
			{ ( props ) => {
				return <NoFollowToggle { ...props } />;
			} }
		</Fill>
	);
}
```

## Screenshots 
screen shot 2019-01-04 at 15 47 14

## Types of changes

New feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/format-library/src/link/inline.js [hunk 1]
```diff
@@ -168,20 +150,71 @@ class InlineLinkUI extends Component {
 		this.setState( { inputValue } );
 	}
 
-	setLinkTarget( opensInNewWindow ) {
-		const { activeAttributes: { url = '' }, value, onChange } = this.props;
-
-		this.setState( { opensInNewWindow } );
+	setLinkAttributes( attributes ) {
+		this.setState( {
+			...this.state,
```
Comment (atimmer) [inline packages/format-library/src/link/inline.js]: You don't have to do this with `setState`. `setState` will already keep the previous state attributes around.

Comment (Xyfi) [reply packages/format-library/src/link/inline.js]: Ah ofcourse. My mistake :).

**Comment by Chrico** [comment]
Howdy,

any news here about the state or if something is missing? Is there already a scheudule when this will be included into Gutenberg? Currently there are a lot of options which could be possible be added to the LinkSettings to support e.G. custom data-attributes for tracking or even kind of theming to open modals.

**Comment by Soean** [comment]
@Xyfi Can you add the No Follow example to the docs?

**Comment by designsimply** [comment]
Noting a related discussion from the #core-editor meeting in WP Slack today: https://wordpress.slack.com/archives/C02QB2JS7/p1547046547434300

File: packages/format-library/README.md [hunk 2]
```diff
@@ -13,3 +13,98 @@ npm install @wordpress/format-library --save
 _This package assumes that your code will run in an **ES2015+** environment. If you're using an environment that has limited or no support for ES2015+ such as lower versions of IE then using [core-js](https://github.com/zloirock/core-js) or [@babel/polyfill](https://babeljs.io/docs/en/next/babel-polyfill) will add support for these methods. Learn more about it in [Babel docs](https://babeljs.io/docs/en/next/caveats)._
 
 <br/><br/><p align="center"><img src="https://s.w.org/style/images/codeispoetry.png?1" alt="Code is Poetry." /></p>
+
+## Extensibility
```
Comment (iseulde) [inline packages/format-library/README.md]: These docs feel like the belong to the link format itself, not to this module in general. Should the read me file be added there?

File: packages/format-library/src/link/inline.js [hunk 3]
```diff
@@ -257,12 +288,24 @@ class InlineLinkUI extends Component {
...
 					focusOnMount={ showInput ? 'firstElement' : false }
+					className="editor-format-toolbar__link-settings-container"
 					renderSettings={ () => (
-						<ToggleControl
-							label={ __( 'Open in New Tab' ) }
-							checked={ opensInNewWindow }
-							onChange={ this.setLinkTarget }
-						/>
+						<div className="editor-format-toolbar__link-settings-container">
+							<ToggleControl
```
Comment (iseulde) [inline packages/format-library/src/link/inline.js]: Can we use our own API here? (Use the Slot we are providing)?

**Comment by GlennMartin1** [comment]
+1 to have this feature ASAP. 

Any ETA? 5.1?

**Comment by mapk** [comment]
I'd like to point to #10128 in an effort to keep the styling and interaction consistent. This is what that interaction looks like.

link-button

**Comment by gziolo** [comment]
> I'd like to point to #10128 in an effort to keep the styling and interaction consistent. This is what that interaction looks like.

This makes me wonder whether we should include more options it in the core by default without adding the overhead of slots and fills. Maybe we should add more options under 3 dots menu based on this PR and Button block. In particular, in the case of the Button block, these Link Settings sections feel super disconnected from the UI. We should think about the flow when someone has to use the keyboard to set link rel after providing url. I think it would be super annoying to tab as much to edit related settings.

**Comment by raquelmsmith** [comment]
Bump and +1 on this. Many, many bloggers rely on the ability to add `rel="nofollow"` to their links for sponsored and affiliate links. We used to add it with our plugin but now we are unable to provide that functionality. I don't have a strong opinion on whether it should be included in WordPress Core, but there at least needs to be the ability to add that feature in a plugin. 

This is an important SEO consideration for all bloggers who use WordPress to make money through any sort of sponsored or affiliate content.

**Comment by danielbachhuber** [comment]
@youknowriad @gziolo Can you summarize what needs to be done in order to land this pull request, or whether there's a suggested workaround to allow `rel=nofollow` attributes to be added with UI?

**Comment by wormeyman** [comment]
Please please land this! @Xyfi is there a way for us to include it now easily on our own sites?

**Comment by youknowriad** [comment]
Due to the precedent in terms of extensibility this PR creates, I'd like to suggest we land the "nofolllow" option as part of the core link modal as a first step.

**Comment by jdcauley** [comment]
What are the odds that this will be added to the core link modal in the near future, it solves a significant problem for a large number of users.

**Comment by designsimply** [comment]
@jdcauley thanks for your interest, generally, unless something is part of an active sprint or milestone, I don't like to predict a time for when something may or may not be able to be finished. In this case, it doesn't look like the PR has been active in several weeks but also it had needed some direction which it now has.

@Xyfi thank you for your work so far! Is this something you have time to follow up on?

<blockquote>land the "nofolllow" option as part of the core link modal as a first step.</blockquote>

**Comment by chrisvanpatten** [comment]
I would love to see this land as an extensible option. I can think of a few cases in our Gutenberg implementation where it would be a much better editorial experience to extend the existing link settings vs having to build a separate link button.

**Comment by mapk** [comment]
@Xyfi can you rebase this PR and bring it up to date? I'd like to have another round of testing here.

**Comment by Xyfi** [comment]
I will work on this in the coming week.

**Comment by Xyfi** [comment]
I rebased the PR on master.

**Comment by mapk** [comment]
I get this error in the browser when trying to test this PR: 

`Notice: Trying to get property 'base' of non-object in /var/www/html/wp-content/plugins/gutenberg/gutenberg.php on line 133 `

**Comment by Xyfi** [comment]
@mapk I didn't touch any PHP, so It's extremely unlikely this is due to my PR. :)

Also I'm not able to reproduce it.

**Comment by mapk** [comment]
I just tested again and the PR is up and running smooth. Thanks! @gziolo or @youknowriad can we get your review on this?

**Comment by youknowriad** [comment]
I think we proposed that we remove the slot for now, can we implement this without the Slot?

**Comment by karmatosed** [comment]
>  these Link Settings sections feel super disconnected from the UI. 

I agree with @gziolo on this. I do wonder if someone is going to see the sidebar in this case.

**Comment by Xyfi** [comment]
@youknowriad @karmatosed I will make time this week to update the PR.

**Comment by chriscct7** [comment]
> This makes me wonder whether we should include more options it in the core by default without adding the overhead of slots and fills. 

There are a lot of options that could/should be there. Some of those are documented on #11599

For example:

- nofollow links
- open in new tab option
- download attribute (for download links)
- ability for plugins to add additional items as relevant (for example if you had a plugin that let you add tags like rel="friend" and things like that

I agree that core should likely add the first 3 as defaults, but a slot is the way to solve this for sure, and notably the lack of extensibility in this space is a huge regression from classic editor to Gutenberg

**Comment by Chrico** [comment]
For me this is the wrong point to add this specific feature. It should be abstracted into a "general thing", which is re-used accross all Components which can have HTML-attributes [1].

There it should allow to add/remove key/value pairs (value could be maybe even a "selection of predefined allowed values"). Also blacklisting "core"-attributes (not changeable) - like on the Link-Component the `href` - should be possible.

---

[1] https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes

**Comment by TwisterMc** [comment]
FYI: There's a plugin called [EditorsKit](https://wordpress.org/plugins/block-options/) that you can install and get nofollow functionality today. Not sure how they're doing it, but you can try it today.

**Comment by mcsf** [comment]
The Link UI has evolved a lot since this PR was in active development. What is the status of the PR?

**Comment by draganescu** [comment]
@Xyfi considering that now the linking mechanism evolved into using `LinkControl` I will close this PR, since given the comments there is not a definitive path ahead and also even if it were the component to update is a completely new one. Thanks for the work it still remains valuable. I would consider  opening an issue about how this extensibility should work? ( cc @Chrico )
